### ISSUE TITLE
Whiten Extremal Value

### ISSUE THREAD

**Issue by garrettwrong** [ComputationalCryoEM/ASPIRE-Python#658]
User reported bug with whitening.  Let's confirm and see if we can incorporate any fixes from the MATLAB code.

```
Hi,

We ran into a strange behavior of the whitening process.
The estimated whitening filter seems to have zero at the center, and so when dividing the filter, you get huge values in the whitened image.
In the MATLAB code I have safeguards against division by small values.
I don’t know if the Python code has implemented them.
Attached is a minimal example demonstrating the problem.
(Rename scripy.txt to script.py, since outlook blocks Python files).
Thanks,
Yoel

```

Added .txt extension to all files for upload.

[example.star.txt](https://github.com/ComputationalCryoEM/ASPIRE-Python/files/9151093/example.star.txt)
[script.py.txt](https://github.com/ComputationalCryoEM/ASPIRE-Python/files/9151094/script.py.txt)
[example.mrcs.txt](https://github.com/ComputationalCryoEM/ASPIRE-Python/files/9151090/example.mrcs.txt)

**Comment by janden** [comment]
It's strange that we get a zero value at the center here. In theory, these filer estimates should all be positive. All the same, might be good to add a small epsilon before dividing.

**Comment by j-c-c** [comment]
Just looking at the noise psd estimation in the `AnisotropicNoiseEstimator()` below. Line 133 sets the center value of the filter. Looks like if `noise_psd_est[mid, mid]` and `mean_est**2` are equal, the center will be zero (unless I'm misunderstanding this). 
https://github.com/ComputationalCryoEM/ASPIRE-Python/blob/1bff8d3884183203bd77695a76bccb1efc909fd3/src/aspire/noise/noise.py#L112-L133

At any rate, in our `whiten()` method we element wise divide by the square root of the noise psd, where as in the [matlab code](https://github.com/PrincetonUniversity/aspire/blob/e697eee5bb185ab1962f1d6e5becaf9caa277732/projections/epsd/cryo_prewhiten.m) the division only takes place if the psd is larger than some threshold value (around lines 66-70), otherwise the values are set to zero. 

If we add some small epsilon before dividing won't that result in a very large value? And would that be problematic?

**Comment by janden** [comment]
> Looks like if `noise_psd_est[mid, mid]` and `mean_est**2` are equal, the center will be zero (unless I'm misunderstanding this).

Right but that is quite unlikely to happen. They are only equal if the mean value of each image is constant (which it may be due to previous preprocessing, i.e. `normalize_bg`). It's worth checking to see if this is the case.

> At any rate, in our `whiten()` method we element wise divide by the square root of the noise psd, where as in the [matlab code](https://github.com/PrincetonUniversity/aspire/blob/e697eee5bb185ab1962f1d6e5becaf9caa277732/projections/epsd/cryo_prewhiten.m) the division only takes place if the psd is larger than some threshold value (around lines 66-70), otherwise the values are set to zero.

Yes that corresponds to a pseudoinverse formulation (since the standard inverse does not exist if one of the filter values is zero).

> If we add some small epsilon before dividing won't that result in a very large value? And would that be problematic?

Adding an epsilon corresponds to regularizing the inverse (as opposed to replacing it with a pseudoinverse). They're two options with different properties. Note that the pseudoinverse also gives large values if the threshold used is very large (i.e., if your threshold is ε and you have the value ε + 0.001, you'll divide by 1 / (ε + 0.001), which is almost as large as 1 / ε. The regularized option has the advantage that it does not introduce discontinuities in the spectrum while the pseudoinverse may be better suited in other situations.

Before we go further on this. it's important to know _why_ we're getting zeros in the filter estimate. It shouldn't really happen under normal circumstances.

**Comment by j-c-c** [comment]
@janden Thanks for the clarification. I'll see if I can track down where these zeros are coming from.

**Comment by j-c-c** [comment]
Ok @janden, I've resumed work on this and I've tracked down where the zeros are coming from. In the `estimate_noise_psd()` method linked in an above comment we take the centered fft of the masked images. In the case of the script that Yoel provided there is an earlier step of `normalize_bg` which causes `images_masked` to have a mean value of zero. In turn, the fft `im_masked_f` has a center value of essentially zero for each image. This then causes the filter to have a center value close to zero.

**Comment by janden** [comment]
Ok got it. Thanks for taking a look at this again.

**Comment by garrettwrong** [comment]
If the only extremal value of concern is actually the center (dc), then we could clamp it with `numpy.nan_to_num`. That would convert an inf to a large finite number.  Assuming the center values are actually 0 the (convolution's) multiply would be 0*large_num. If that works, a reaonably clean implementation would be to use a context for the application of the power and subsequent nan conversion.  @j-c-c , let me know if you think that would work; I expect you'll fill me in later. Thanks

**Comment by j-c-c** [comment]
#1025 added a safeguard against extremal values for power filters with negative power. Closing.

**Comment by garrettwrong** [comment]
Reopening.  It seems something like this may be coming up again after #1125 .  I suspect we are running/evaluating something filter related in singles now and that is causing a bit of trouble in practice that we are not covering in unit tests...

Specifically, I found running the `gallery/experiments/simulated_abinitio_pipeline.py` in single precision fails.  The resulting preprocessed images look like they are junk. (Example has been changed to run as doubles for v0.13.0).

#

Here are the gallery sim images, noise filter and preproc images with doubles:


<img width="1667" alt="Screenshot 2024-10-08 at 11 29 19 AM" src="https://github.com/user-attachments/assets/ccbaf75f-9497-4f78-90ef-ef7714074781">
<img width="558" alt="Screenshot 2024-10-08 at 11 29 27 AM" src="https://github.com/user-attachments/assets/3ae05ea8-279b-43bc-aa38-66764e05fa48">
<img width="1588" alt="Screenshot 2024-10-08 at 11 29 36 AM" src="https://github.com/user-attachments/assets/2ad170af-b78e-4010-8ad8-31d2e2731ecc">


#

and the same, as singles...  Note the last log line:

```
(v0.13.0py39) ➜  experiments git:(simgalhotpatch) ✗ python simulated_abinitio_pipeline.py               
2024-10-08 11:30:10,083 INFO [aspire.volume.volume] /Users/gbwright/Library/Caches/ASPIRE-data/emdb_2660.map with dtype float32 loaded as None
2024-10-08 11:30:10,084 INFO [__main__] Original volume map data shape: (1, 360, 360, 360) dtype:float32
2024-10-08 11:30:10,084 INFO [__main__] Initialize CTF filters.
2024-10-08 11:30:10,087 INFO [aspire.source.image] Creating Simulation with 10000 images.
2024-10-08 11:30:10,110 INFO [aspire.source.simulation] Appending CustomNoiseAdder to generation pipeline
2024-10-08 11:30:10,143 INFO [aspire.source.image] Setting max. resolution of source = 32
2024-10-08 11:30:10,241 INFO [aspire.source.image] Caching source images
2024-10-08 11:45:44,057 INFO [__main__] Perform phase flip to input images.
2024-10-08 11:45:44,600 INFO [aspire.source.image] Perform phase flip on source object
2024-10-08 11:45:44,600 INFO [aspire.source.image] Adding Phase Flip Xform to end of generation pipeline
100%|████████████████████████████████████████████████████████████████████████████| 20/20 [00:00<00:00, 28.84it/s]
2024-10-08 11:45:45,478 INFO [aspire.source.image] Whitening source object
2024-10-08 11:45:45,478 INFO [aspire.source.image] Transforming all CTF Filters into Multiplicative Filters
2024-10-08 11:45:45,479 INFO [aspire.source.image] Adding Whitening Filter Xform to end of generation pipeline
2024-10-08 11:45:52,232 WARNING [aspire.operators.filters] PowerFilter setting 1024 extremal filter value(s) to zero.
```
<img width="1562" alt="Screenshot 2024-10-08 at 11 45 42 AM" src="https://github.com/user-attachments/assets/8309e860-ddd9-4e3e-a814-4f7235f7f5e0">
<img width="518" alt="Screenshot 2024-10-08 at 11 45 50 AM" src="https://github.com/user-attachments/assets/64dc8dc2-103a-4ab6-9df2-6c283a0a6444">
<img width="1602" alt="Screenshot 2024-10-08 at 11 46 02 AM" src="https://github.com/user-attachments/assets/8a4dd919-4aa7-479f-bf87-c4a56863877e">

**Comment by garrettwrong** [comment]
Closed by #1195
